AudioDeviceExample/AudioDevices/ExampleAudioEngineDevice.m

```diff
@@ -0,0 +1,741 @@
+            [self setupAVAudioSession];
+            _rendererInitialized = YES;
+        }
+
+        /*
+         * Assume that the AVAudioSession has already been configured and started and that the values
+         * for sampleRate and IOBufferDuration are final.
+         */
+        _renderingFormat = [[self class] activeFormat];
+    }
+
+    return _renderingFormat;
+}
+
+- (BOOL)initializeRenderer {
+    /*
+     * In this example we don't need any fixed size buffers or other pre-allocated resources. We will simply write
+     * directly to the AudioBufferList provided in the AudioUnit's rendering callback.
+     */
+    return YES;
+}
+
+- (BOOL)startRendering:(nonnull TVIAudioDeviceContext)context {
+    @synchronized(self) {
+        NSAssert(self.renderingContext == NULL, @"Should not have any rendering context.");
+
+        self.renderingContext = malloc(sizeof(RendererAudioContext));
+        memset(self.renderingContext, 0, sizeof(RendererAudioContext));
+
+        self.renderingContext->deviceContext = context;
+        self.renderingContext->maxFramesPerBuffer = _renderingFormat.framesPerBuffer;
+        if (@available(iOS 11.0, *)) {
+            self.renderingContext->renderBlock = (__bridge void *)(_engine.manualRenderingBlock);
+        }
+
+        [self setupAudioEngine];
+
+        if (self.capturingContext) {
+            if (![self setupAudioUnitWithRenderContext:self.renderingContext
+                                        capturingContext:self.capturingContext]) {
+                free(self.renderingContext);
+                self.renderingContext = NULL;
+                return NO;
+            }
+            return [self startAudioUnit];
+        }
+    }
+    return YES;
+}
+
+- (BOOL)stopRendering {
+    _rendererInitialized = NO;
+    [self stopAudioUnit];
+
+    @synchronized(self) {
+        [self teardownAudioUnit];
+        [self teardownAudioEngine];
+
+        NSAssert(self.renderingContext != NULL, @"Should have a rendering context.");
+        free(self.renderingContext);
+        self.renderingContext = NULL;
+
+        NSAssert(self.capturingContext != NULL, @"Should have a capture context.");
+        free(self.capturingContext);
+        self.capturingContext = NULL;
+    }
+
+    return YES;
+}
+
+#pragma mark - TVIAudioDeviceCapturer
+
+- (nullable TVIAudioFormat *)captureFormat {
+    if (!_capturingFormat) {
+
+        if (!self.isRendererInitialized) {
+            [self setupAVAudioSession];
+            _capturerInitialized = YES;
+        }
+
+        /*
+         * Assume that the AVAudioSession has already been configured and started and that the values
+         * for sampleRate and IOBufferDuration are final.
+         */
+        _capturingFormat = [[self class] activeFormat];
+    }
+
+    return _capturingFormat;
+}
+
+- (BOOL)initializeCapturer {
+    _audioBufferList.mNumberBuffers = 1;
+    _audioBufferList.mBuffers[0].mNumberChannels = kPreferredNumberOfChannels;
```

Comment:
Do we need to fill out `mDataByteSize` as well?

Reply:
Its OK not to set it because it gets set by the AudioUnit. 

BTW I had used this code example - https://github.com/GoogleCloudPlatform/ios-docs-samples/blob/cda5b8089fb741b3e54ef83dca9d4bfe7000c47a/speech/Objective-C/Speech-gRPC-Streaming/Speech/AudioController.m#L77